### PR TITLE
feat: Lazy engine initialization

### LINKED ISSUE #2673: ✨[Feature] Delayed Initialization for `TRTModule` Classes
# Context
For models requiring fallback to Torch due to converter capabilities, custom operators, or other needs, each of the `TRTEngine` objects is compiled, initialized, inserted into the Torch `nn.Module`, and runtime-ready during compile time. This takes up an unnecessary amount of memory on the GPU at compile time.

# Proposal
Use the GPU as a build space for `TRTEngine` objects, but do not deserialize or initialize the engines until the first forward pass, similar to what is done here:
https://github.com/pytorch/TensorRT/blob/ad74a735056667726692c49a175a790647ef889e/py/torch_tensorrt/fx/trt_module.py#L25-L27

## API Details
The `TRTModule` objects will take a parameter, `construct_live=True`, which can be specified to `False` if it is desired to initialize the engines at the first `forward` pass, thereby avoiding unnecessary usage of GPU space during compilation. After building the engine at compile time, the serialized object is moved to host memory until runtime, at which point it is initialized. `check_initialized()` is called at every `forward` pass, only having a measurable effect on the first pass of inference at which point the engines are moved from host to device memory for usage.

### PR DESCRIPTION
**Pull request by narendasan**
# Description

Allows engines to not be setup immediately after compilation but all at once before the module is returned back to the user.

Fixes #2673 

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 1]
```diff
@@ -129,17 +140,13 @@ def _load_from_state_dict(
...
         # Run multi-gpu device check to validate engine instantiation
         multi_gpu_device_check()
-
-        runtime = trt.Runtime(TRT_LOGGER)
-        self.engine = runtime.deserialize_cuda_engine(engine_bytes)
-
-        self.input_names = state_dict[prefix + "input_names"]
-        self.output_names = state_dict[prefix + "output_names"]
-        self._initialize()
+        self.setup_engine()
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:149]: Add more tests to verify serialization behavior with lazy modules

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py [hunk 2]
```diff
@@ -50,13 +50,16 @@ class TorchTensorRTModule(torch.nn.Module):  # type: ignore[misc]
         output_binding_names (List[str]): List of output TensorRT engine binding names in the order they should be returned
     """
 
+    defer_engine_setup = False
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py:53]: Remove

**Review by peri044 — COMMENTED** [review]
LGTM. Could you add a testcase with a model that has fallback ops ?

**Review by zewenli98 — APPROVED** [review]
LGTM

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 3]
```diff
@@ -30,19 +30,49 @@ class PythonTorchTensorRTModule(Module):  # type: ignore[misc]
     FX / Dynamo / Python deployments. This module cannot be serialized to torchscript via torch.jit.trace for C++ deployment.
     """
 
+    defer_engine_setup = False
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:33]: Is it used below?

**Review by zewenli98 — COMMENTED** [review]
LGTM, just a few minor comments

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 4]
```diff
@@ -358,9 +359,11 @@ def run(
...
         )
 
-        return TRTInterpreterResult(
-            serialized_engine, self._input_names, self._output_names
-        )
+        with io.BytesIO() as engine_bytes:
+            engine_bytes.write(serialized_engine)
+            engine_str = engine_bytes.getvalue()
+
+        return TRTInterpreterResult(engine_str, self._input_names, self._output_names)
```
[the remark this answers is not in the document]
Comment (narendasan) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:366]: because its not actually a byte array, its a buffer

File: py/torch_tensorrt/dynamo/_compiler.py [hunk 5]
```diff
@@ -647,10 +652,5 @@ def convert_module_to_trt_engine(
...
 
-    import io
-
-    with io.BytesIO() as engine_bytes:
-        engine_bytes.write(interpreter_result.engine)
-        engine_bytearray: bytes = engine_bytes.getvalue()
-
-    return engine_bytearray
+    serialized_engine: bytes = interpreter_result.serialized_engine
+    return serialized_engine
```
[the remark this answers is not in the document]
Comment (narendasan) [reply py/torch_tensorrt/dynamo/_compiler.py:656]: i think its okay to leave as is

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 6]
```diff
@@ -32,17 +32,45 @@ class PythonTorchTensorRTModule(Module):  # type: ignore[misc]
...
+        output_binding_names: Optional[List[str]] = None,
+        *,
+        name: str = "",
         settings: CompilationSettings = CompilationSettings(),
     ):
+        """Takes a name, target device, serialized TensorRT engine, and binding names / order and constructs
+        a PyTorch ``torch.nn.Module`` around it. Uses TensorRT Python APIs to run the engine
+
+        Arguments:
+            serialized_engine (bytearray): Serialized TensorRT engine in the form of a bytearray
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:46]: Are `bytearray` and `bytes` interchangeable?

Comment (narendasan) [reply py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:46]: yeah the difference is mutability
